libcudacxx/include/cuda/std/__utility/typeid.h

```diff
@@ -0,0 +1,507 @@
+}
+
+// A quick smoke test to ensure that the pretty name extraction is working.
+static_assert(3 == __string_view("int").size(), "3 == __string_view(int).size()");
+static_assert(3 == __pretty_nameof<int>().size(), "3 == __pretty_nameof<int>().size()");
+static_assert(__pretty_nameof<int>() == __string_view("int"), "__pretty_nameof<int>() == __string_view(\"int\")");
+static_assert(__pretty_nameof<float>() < __pretty_nameof<int>(), "__pretty_nameof<float>() < __pretty_nameof<int>()");
+
+/// @brief A minimal implementation of `std::type_info` for platforms that do
+/// not support RTTI.
+struct __type_info
+{
+  __type_info()                   = delete;
+  __type_info(__type_info const&) = delete;
+
+  _LIBCUDACXX_HIDE_FROM_ABI explicit constexpr __type_info(__string_view __name) noexcept
+      : __name_(__name)
+  {}
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr char const* name() const noexcept
+  {
+    return __name_.begin();
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr __string_view __name_view() const noexcept
+  {
+    return __name_;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr bool before(const __type_info& __other) const noexcept
+  {
+    return __name_ < __other.__name_;
+  }
+
+  // Not yet implemented:
+  // _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr size_t hash_code() const noexcept
+  // {
+  //   return ;
+  // }
+
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator==(const __type_info& __lhs, const __type_info& __rhs) noexcept
+  {
+    return &__lhs == &__rhs || __lhs.__name_ == __rhs.__name_;
+  }
+
+#if _CCCL_STD_VER <= 2017
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator!=(const __type_info& __lhs, const __type_info& __rhs) noexcept
+  {
+    return !(__lhs == __rhs);
+  }
+#endif
+
+private:
+  __string_view __name_;
+};
+
+#if !defined(_CCCL_NO_INLINE_VARIABLES)
+
+template <class _Tp>
+inline constexpr __type_info __typeid{__pretty_nameof<_Tp>()};
+
+#  define _CCCL_CONSTEXPR_TYPEID(...) _CUDA_VSTD::__typeid<_CUDA_VSTD::__remove_cv_t<__VA_ARGS__>>
+
+#else
+
+template <class _Tp>
+struct __typeid_value
+{
+  static constexpr __type_info value{__pretty_nameof<_Tp>()};
+};
+
+// Before the addition of inline variables, it was necessary to
+// provide a definition for constexpr class static data members.
+template <class _Tp>
+constexpr __type_info __typeid_value<_Tp>::value;
+
+#  ifndef _CCCL_NO_VARIABLE_TEMPLATES
+
+template <class _Tp>
+constexpr __type_info const& __typeid = __typeid_value<_Tp>::value;
+
+#    define _CCCL_CONSTEXPR_TYPEID(...) _CUDA_VSTD::__typeid<_CUDA_VSTD::__remove_cv_t<__VA_ARGS__>>
+
+#  else // !_CCCL_NO_VARIABLE_TEMPLATES
+
+template <class _Tp>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr __type_info const& __typeid() noexcept
+{
+  return __typeid_value<_Tp>::value;
+}
+
+#    define _CCCL_CONSTEXPR_TYPEID(...) _CUDA_VSTD::__typeid<_CUDA_VSTD::__remove_cv_t<__VA_ARGS__>>()
```

Comment:
We are within cuda::std:: so qualifying a type is not needed
```suggestion
#    define _CCCL_CONSTEXPR_TYPEID(...) _CUDA_VSTD::__typeid<__remove_cv_t<__VA_ARGS__>>()
```

Reply:
This is a macro definition. We don't know the context in which it will be expanded.